src/experimental/collection_new_iterable_model/shared/iterator/impl/RedBlackTreeIterator.hpp

```diff
@@ -0,0 +1,699 @@
+
+
+        template < typename __ElementType >                                               // NOLINT(bugprone-reserved-identifier)
+        RedBlackTreeForwardIterator < __ElementType > :: RedBlackTreeForwardIterator (    // NOLINT(bugprone-reserved-identifier)
+                RBTreeNode  *  pRoot,
+                RBTreeNode  *  pCurrentNode
+        ) noexcept :
+                AbstractTreeIterator < __ElementType > (
+                        pRoot,
+                        pCurrentNode
+                ) {
+
+        }
+
+
+        template < typename __ElementType >                                             // NOLINT(bugprone-reserved-identifier)
+        RedBlackTreeForwardIterator < __ElementType > :: RedBlackTreeForwardIterator (  // NOLINT(bugprone-reserved-identifier)
+                const RedBlackTreeForwardIterator & iterator
+        ) noexcept :
+                AbstractTreeIterator < __ElementType > ( iterator ) {
+
+        }
+
+
+        template < typename __ElementType >                                             // NOLINT(bugprone-reserved-identifier)
+        RedBlackTreeForwardIterator < __ElementType > :: RedBlackTreeForwardIterator (  // NOLINT(bugprone-reserved-identifier)
+                RedBlackTreeForwardIterator && iterator
+        ) noexcept :
+                AbstractTreeIterator < __ElementType > ( std :: move ( iterator ) ) {
+
+        }
+
+
+        template < typename __ElementType >                                 // NOLINT(bugprone-reserved-identifier)
+        auto RedBlackTreeForwardIterator < __ElementType > :: operator = (  // NOLINT(bugprone-reserved-identifier)
+                RedBlackTreeForwardIterator const & iterator
+        ) noexcept -> RedBlackTreeForwardIterator & {
+
+            if ( this == & iterator ) {
+                return * this;
+            }
+
+            this->_pRoot        = iterator._pRoot;
+            this->_pCurrentNode = iterator._pCurrentNode;
+
+            return * this;
+        }
+
+
+        template < typename __ElementType >                                 // NOLINT(bugprone-reserved-identifier)
+        auto RedBlackTreeForwardIterator < __ElementType > :: operator = (  // NOLINT(bugprone-reserved-identifier)
+                RedBlackTreeForwardIterator && iterator
+        ) noexcept -> RedBlackTreeForwardIterator & {
+
+            if ( this == & iterator )
+                return * this;
+
+            this->_pRoot        = cds :: exchange ( iterator._pRoot, nullptr );
+            this->_pCurrentNode = cds :: exchange ( iterator._pCurrentNode, nullptr );
+
+            return * this;
+        }
+
+
+        template < typename __ElementType >                                 // NOLINT(bugprone-reserved-identifier)
+        auto RedBlackTreeForwardIterator < __ElementType > :: operator ++ ( // NOLINT(bugprone-reserved-identifier)
+        ) noexcept -> RedBlackTreeForwardIterator & {
+
+            this = this->successor;
```

Comment:
decltype (this) = RedBlackTreeForwardIterator < __ElementType > * const. I highly suspect this is not legal. Also lack of fn call.
Consider advance instead of assignation with successor
```cpp
this->advance();
```